File: customer-address-jpa/src/test/java/com/izertis/example/repository/jpa/inmemory/CustomerRepositoryInMemory.java
package com.izertis.example.repository.jpa.inmemory;

import com.izertis.example.domain.Customer;
import com.izertis.example.repository.jpa.CustomerRepository;

import static org.apache.commons.lang3.ObjectUtils.firstNonNull;

public class CustomerRepositoryInMemory extends InMemoryJpaRepository<Customer> implements CustomerRepository {

    @Override
    public <T extends Customer> T save(T entity) {
        Customer finalEntity = super.save(entity);
        finalEntity.getPaymentMethods().forEach(paymentMethod -> {
            paymentMethod.setId(firstNonNull(paymentMethod.getId(), finalEntity.getId()));
        });
        return (T) finalEntity;
    }
}

File: kustomer-address-jpa/src/test/kotlin/com/izertis/example/repository/jpa/inmemory/CustomerRepositoryInMemory.kt
package com.izertis.example.repository.jpa.inmemory

import com.izertis.example.domain.Customer
import com.izertis.example.domain.PaymentMethod
import com.izertis.example.repository.jpa.CustomerRepository
import org.apache.commons.lang3.ObjectUtils
import java.util.function.Consumer

class CustomerRepositoryInMemory : InMemoryJpaRepository<Customer, Long>(), CustomerRepository {

    override fun <S : Customer> save(entity: S): S {
        super.save(entity)
        entity.paymentMethods.forEach { paymentMethod ->
            paymentMethod.id = paymentMethod.id ?: entity.id
        }
        return entity
    }
}
